serverless/components
# [WIP] serverless.yml and state.json encryption

## What has been implemented?

Closes #238 

Example how the encryption could be implemented.

## Steps to verify

Checkout the utils repo PR (https://github.com/serverless/utils/pull/1) and `npm link` it.

Checkout this PR and use it as you dev version of the components. `npm link @serverless/utils` to use the local version of utils that supports yaml options.

Clone the https://github.com/laardee/components-encryption repo (or copy files)

Run `components encrypt --variable some-secret-value` (using you local dev version of components which have the `@serverless/utils` linked with npm link)

Now you should have `.components-credentials` file in the project root. And in the command line should be the encrypted value of the `some-secret-value`. 

Replace the function echo's environmental variable `some-value` with encrypted value

```yaml
 ....
      environment:
        someValue: !encrypted INSERT_VALUE_HERE
```

Now when you run `components deploy`; the encrypted value should be decrypted at the runtime and passed to the lambda as plain text. Also in the state.json the `someValue` should be decrypted.

```json
{
....
  "components-encryption:functionEcho": {
....
      "environment": {
        "someValue": "some-secret-value"
      }
....
```

To encrypt the state.json, so that it can be committed to git, run `components encrypt --state`.

## Todos:
